Dashboard: Manage Locations

sagzy opened the pull request.

shafy commented:
Dude, this looks and feels damn nice! I played around with it a bit and here are my comments. I'll review the code in the next days, but I figured we can already discuss my feedback in the meantime:


Major comments:
- After adding a location, I would like it to zoom in to that location so that I can double check if it's really correctly added (at least I got that urge after adding some locations)
- Adding an address and then changing some of the info (e.g. the zip) still shows the pin at the initial location, but the saved info is wrong. There are several ways we can handle this: 1) show an error if there's no match for the exact address (e.g. wrong zip), 2) use autocomplete, 3) don't care for now, 4) something else?
- Changing the lat or long values doesn't move the pin in the map, so as an user I'm not sure if it worked or not and where the pin now is displayed.
- I would put lat and long fields into separate lines so that we can see all digits or select the whole number on first click (like the url bar in browsers) (it's a bit annoying to edit otherwise)

Minor comments:
- I suggest to remove the big image in the log in and register screens and for now simply center the log in form without the image. I think these kind of images make it look like all other website and very corporate-y. 
- We can also remove the root `/` route, since this will point to our static website.
- Should we remove the search bar in this view? Users might think that they can search their locations and are then disappointed when it doesn't work. Later, we can replace it with an actual search bar that helps users find their locations
- I also noticed that the default zoom of the geolocation feature is not large enough for me. But I can adjust that in a different PR.


I think we also should have a list view of the locations, otherwise it's a pain to find and edit a location. However, we can do that after launch.

What do you think?

shafy commented on app/views/dashboard/locations/show.html.erb:
```diff
@@ -1 +1,25 @@
+  <div class="w-full h-full p-8 flex flex-col">
+    <div class="relative w-full">
+      <div class="flex flex-row justify-between mb-8">
+        <button class="text-left pointer text-sm underline py-1">
+          <%= link_to "Edit", edit_dashboard_map_location_path(map_id: @map.id, id: @location.id), target: '_top' %>
+        </button>
```
Don't you think it's better to just format the `<a>` as a button instead of wrapping it in a button?

shafy commented on app/views/dashboard/maps/_bottom_menu.html.erb:
```diff
@@ -0,0 +1,9 @@
+<div class="flex flex-row justify-center m-4">
+  <button class="button--bottom-menu w-1/6 mx-4 p-4 text-base font-semibold">
+    <%= link_to "📍 Add Location", new_dashboard_map_location_path(@map.id ) %>
+  </button>
```
Same here with the button wrapping.

shafy commented on app/views/development/design/index.html.erb:
```diff
@@ -34,13 +27,12 @@
-  </div>
+
+<div class="flex-auto flex space-x-6">
+  <button class="w-1/6 button--main">
+    <%= link_to "Main option" %>
+  </button>
+  <button class="w-1/6 button--secondary">
+    <%= link_to "Secondary option" %>
+  </button>
```
Button wrapping no good

shafy commented on app/views/home/index.html.erb:
```diff
@@ -1,19 +1,17 @@
-        </button>
-      </div>
+    <div class="flex justify-center my-8 space-x-6">
+      <button class="w-1/4 button--main" type="button">
+        <%= link_to "Get started", new_user_registration_path %>
+      </button>
+      <button class="w-1/4 button--secondary" type="button">
+        <%= link_to "Log in", new_user_session_path %>
+      </button>
```
Button wrapping

shafy commented on spec/factories/locations.rb:
```diff
@@ -29,14 +28,20 @@
+    end
+
+    description {
+      'Founded by King Aegon I Targaryen, King of the Andals and the First Men.'
```
All hail King Aegon I!

shafy reviewed (commented):
Great stuff! Nothing major, just a few things that might improve the code - but everything optional from my side :-)

sagzy commented on app/controllers/concerns/MapsControllable.rb:
```diff
@@ -1,23 +0,0 @@
-# frozen_string_literal: true
```
For modularity reasons. e.g. now we can access @map.bounds wherever needed in the code. 

Also, the dashboord map controller & the public map controller do not necessarily need the same variable assignments in their show action, so I'd rather have more flexibility there

shafy replied:
Sounds good!

sagzy commented on app/controllers/dashboard/locations_controller.rb:
```diff
@@ -2,12 +2,61 @@
+    end
+
+    def edit
+      @bounds = @map.bounds
```
True

sagzy commented on app/controllers/dashboard/maps_controller.rb:
```diff
@@ -2,14 +2,18 @@
+      @map = Map.find(params[:id])
+      @bounds = @map.bounds
+      @markers_json = @map.markers.to_json
+      @location_base_url = "/dashboard/maps/#{@map.id}/locations/"
```
Yes ideally, but it's not the full url here (the `/:id`is missing at the end)`.

This base url is completed in the map controller JS, when we create the markers (each marker links to the show action of the corresponding location)

shafy replied:
I see. What if you use the path helper with a random `:location_id`, like 1, and then remove the last character? Something like:
```
dashboard_maps_locations_path(@map, 1).delete_suffix(1)
```

Too unclear? Too hacky?

sagzy replied:
good idea! thanks

sagzy commented on app/controllers/dashboard/maps_controller.rb:
```diff
@@ -2,14 +2,18 @@
+      @bounds = @map.bounds
+      @markers_json = @map.markers.to_json
+      @location_base_url = "/dashboard/maps/#{@map.id}/locations/"
+      @ask_location_permission = false
```
Thanks for the feedback dude! Glad you like it and very good points raised. Some follow up questions on two points:

> Changing the lat or long values doesn't move the pin in the map, so as an user I'm not sure if it worked or not and where the pin now is displayed.

Ah good to know that you used it this way. I actually had in mind that the user will drag the marker, and not add lat / long inputs manually. In that case, I should disable the lat / longitude fields. Do you see a use case where the user would prefer to enter lat / long values instead of dragging the marker?

>  Users might think that they can search their locations and are then disappointed when it doesn't work. Later, we can replace it with an actual search bar that helps users find their locations

Yeah good idea; mapbox geocoding can actually be superpowered to also show our locations as results (see [example](https://docs.mapbox.com/mapbox-gl-js/example/forward-geocode-custom-data/)). This could be a nice addition, it doesn't seem to complicated to add, I'll see if I can add it quickly. It will then also solve the "show a list of locations in my dashboard" problem.

shafy replied:
To be clear, I was referring to the  `@ask_location_permission = false` assignment.

sagzy replied:
We could handle that outside of the controller, but then you end up with logic in the view (if action is x then do y, etc.), which I don't find great either. The reason why I have to add these variables here is because I've tried to dry up the view (share the same between dashboard & public) & I think when we need to set too many variables, it's a good sign to un-dry the view & slip it into two separate views

shafy replied:
Ok, up to you. Just saying that some logic in the view is ok, and there's a reason the `[current_page?](https://apidock.com/rails/ActionView/Helpers/UrlHelper/current_page%3f)` view helper exists. Often, just assining bool variables in controller to be used in view is also code smell. As I said, up to you. I can live with it :-)

sagzy commented:
Thanks for the feedback dude! Glad you like it and very good points raised. Some follow up questions on two points:

> Changing the lat or long values doesn't move the pin in the map, so as an user I'm not sure if it worked or not and where the pin now is displayed.

Ah good to know that you used it this way. I actually had in mind that the user will drag the marker, and not add lat / long inputs manually. In that case, I should disable the lat / longitude fields. Do you see a use case where the user would prefer to enter lat / long values instead of dragging the marker?

>  Users might think that they can search their locations and are then disappointed when it doesn't work. Later, we can replace it with an actual search bar that helps users find their locations

Yeah good idea; mapbox geocoding can actually be superpowered to also show our locations as results (see [example](https://docs.mapbox.com/mapbox-gl-js/example/forward-geocode-custom-data/)). This could be a nice addition, it doesn't seem to complicated to add, I'll see if I can add it quickly. It will then also solve the "show a list of locations in my dashboard" problem.

shafy commented:
> Ah good to know that you used it this way. I actually had in mind that the user will drag the marker, and not add lat / long inputs manually. In that case, I should disable the lat / longitude fields. Do you see a use case where the user would prefer to enter lat / long values instead of dragging the marker?

Good point - let's disable them for now (I mean there might be use cases, but let's keep it simple!)

> Yeah good idea; mapbox geocoding can actually be superpowered to also show our locations as results (see example). This could be a nice addition, it doesn't seem to complicated to add, I'll see if I can add it quickly. It will then also solve the "show a list of locations in my dashboard" problem.

Good point. Ok, you can look into it but my feeling is that it's not necessary for launch, so we shouldn't invest extra time there. I still think a list of location can be useful to quickly go through your locations, find stuff and so on. But not required for launch.

sagzy commented:
> Good point. Ok, you can look into it but my feeling is that it's not necessary for launch, so we shouldn't invest extra time there. I still think a list of location can be useful to quickly go through your locations, find stuff and so on. But not required for launch.

Yes agree, not essential for the launch. But I think the list of location could be merged with the search, to simplify the interface (rather than having a new view with all the locations). I think it goes in line with the WISIWYG approach: minimum number of screens and do a maximum of things directly on the map. The search bar could render a list of all locations by default when you click on it, then narrow down based on the user input until one location can be selected

shafy commented:
I think a separate list view fulfills a different need to a map view. Let's say I have 30 locations and I want to edit one. With the map-only approach, I need to start typing in the name and select it from the list or zoom into where it is and click edit. I personally would prefer a list view for this, since it's much quicker to scroll through and select the location you want to edit. Often times, I don't care how my map looks like, but I rather want to find something quickly. Loading the map is always less snappy and more distracting than loading a list of locations.

WYSIWYG is fine, but not always the best solution I think.

But we can discuss this more in the future, since we both agree we don't need it for launch :-)

sagzy commented:
> Let's say I have 30 locations and I want to edit one. With the map-only approach, I need to start typing in the name and select it from the list or zoom into where it is and click edit. I personally would prefer a list view for this, since it's much quicker to scroll through and select the location you want to edit. 

Fair enough. What if you now have 100 locations? Or 500? You would prefer some sort of search & find feature then rather than just scroll a long list, right? I think it can be essentially the same thing with the search bar, you click on the search bar and you see a "view list" of all available locations (eventually narrowed down when it's not possible to show all of them), and then you start to type and find the location you want to edit quickly. If you don't want to search, just scroll and find it.

I agree with you that WYSIWYG is not going to be enough for every feature. Let's say we want to enable bulk addition / edition of locations, then you def. need a separate view outside of the map. But in general, I think we should try to keep the dashboard interface as light as possible, i.e. favor not adding a new view if we can offer a good UX without it.

> But we can discuss this more in the future, since we both agree we don't need it for launch :-)

Deal. I stop now :D

sagzy commented:
Addressed the code comments now! And added the product comments in basecamp to tackle them in due time. I'll merge that branch to main and then you can continue on top @shafy?

shafy commented:
Sounds good to me, thanks @sagzy !
